### PR TITLE
Redis Queue

### PR DESCRIPTION
**Pull request by sdlyy**
Resolves L2B-5387

Check out readme on how to run queue-based scenario.
The main standalone scenario is still in play.

Most up to date excalidraw link:
https://app.excalidraw.com/s/1Pobo8fNXle/22icgOkmpSn

Batteries included:
- Queue setup wrapper
- Queue setup with processor at once
- Routing (forward, broadcast, by routing key)
- Worker retries and logging
- Dashboard if port for dashboard has been provided
- Basic type-safety and payload declaration
- Complete setup as `queue.ts`

Feel free to mess around and setup your own flows.
If something goes wrong, re-seed the database and flush the redis (see README)

image

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,26 +12,33 @@
...
+    "format:fix": "biome check --apply .",
     "lint": "biome lint .",
     "lint:fix": "biome lint . --apply",
     "nanoid": "nanoid",
     "seed": "tsx src/db/seed.ts",
     "start": "tsx src/main.ts",
     "test": "vitest",
-    "typecheck": "tsc --noEmit"
+    "typecheck": "tsc --noEmit",
+    "queue": "tsx --no-warnings=ExperimentalWarning --watch src/queue.ts"
```
Comment (imxeno) [inline package.json]: If it's a dev script I guess it should be called differently

File: src/sources/tokenList.ts [hunk 2]
```diff
@@ -37,7 +40,7 @@ function buildTokenListSource({ db, url, tag, logger }: Dependencies) {
 
       return {
         networkId: chain.id,
-        address: token.address.toUpperCase(),
+        address: checksumAddress(token.address as `0x${string}`),
```
Comment (imxeno) [inline src/sources/tokenList.ts]: Can't we use viem's `getAddress` here?

File: src/utils/queue/wrap.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { Queue } from 'bullmq'
+
+export type TokenUpdateQueue = ReturnType<typeof wrapTokenQueue>
+export type DeploymentUpdatedQueue = ReturnType<
+  typeof wrapDeploymentUpdatedQueue
+>
+
+// Thin abstraction over the queue so we explicitly assign the job id to
```
Comment (imxeno) [inline src/utils/queue/wrap.ts]: If this applies to both function (which I guess it does), we should leabe this commend also on `wrapDeploymentUpdatedQueue`

File: src/queue.ts [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+      tag,
+      url,
+      logger,
+      db,
+      queue: tokenUpdateQueue,
+    }),
+  }),
+)
+
+// const lzV1Sources = networksConfig.filter(withExplorer).map((networkConfig) => {
```
Comment (imxeno) [inline src/queue.ts:269]: Is layerzero disabled on purpose?

Comment (sdlyy) [reply src/queue.ts:269]: Yup

File: src/utils/queue/wrap.ts (hunk 3 above)
Comment (sdlyy) [reply src/utils/queue/wrap.ts]: https://github.com/l2beat/token-db/pull/31/commits/14331efa99390b178f85dd98055e2119019033ee

File: src/sources/tokenList.ts (hunk 2 above)
Comment (sdlyy) [reply src/sources/tokenList.ts]: be my guest
https://github.com/l2beat/token-db/pull/31/commits/3fdd751b81be81e4bab5ac233e5c5e79bfc039d3

File: package.json (hunk 1 above)
Comment (sdlyy) [reply package.json]: no idea tbh, I'm terrible at naming things and invalidating cache
https://github.com/l2beat/token-db/pull/31/commits/f64886124f70b8adfae58af550a881f2ca9d5c92

**Review by imxeno — APPROVED** [review]

**Pull request merged**